source/nRF5xCharacteristicDescriptorDiscoverer.cpp

```diff
@@ -0,0 +1,208 @@
+
+namespace { 
+    void emptyDiscoveryCallback(const CharacteristicDescriptorDiscovery::DiscoveryCallbackParams_t*) { }
+    void emptyTerminationCallback(const CharacteristicDescriptorDiscovery::TerminationCallbackParams_t*) { }
+}
+
+nRF5xCharacteristicDescriptorDiscoverer::nRF5xCharacteristicDescriptorDiscoverer(size_t concurrentConnectionsCount) : 
+    maximumConcurrentConnectionsCount(concurrentConnectionsCount), 
+    discoveryRunning(new Discovery[concurrentConnectionsCount]) {
+
+}
+
+nRF5xCharacteristicDescriptorDiscoverer::~nRF5xCharacteristicDescriptorDiscoverer() { 
+    delete [] discoveryRunning;
+}
+
+ble_error_t nRF5xCharacteristicDescriptorDiscoverer::launch(
+    const DiscoveredCharacteristic& characteristic, 
+    const CharacteristicDescriptorDiscovery::DiscoveryCallback_t& discoveryCallback,
+    const CharacteristicDescriptorDiscovery::TerminationCallback_t& terminationCallback
+) {
+    Gap::Handle_t connHandle = characteristic.getConnectionHandle();
+    // it is ok to deduce that the start handle for descriptors is after 
+    // the characteristic declaration and the characteristic value declaration
+    // see BLUETOOTH SPECIFICATION Version 4.2 [Vol 3, Part G] (3.3)
+    Gap::Handle_t descriptorStartHandle = characteristic.getDeclHandle() + 2;
+    Gap::Handle_t descriptorEndHandle = characteristic.getLastHandle();
+
+    // check if their is any descriptor to discover
+    if (descriptorEndHandle < descriptorStartHandle) { 
+        CharacteristicDescriptorDiscovery::TerminationCallbackParams_t termParams = { 
+            characteristic,
+            BLE_ERROR_NONE
+        };
+        terminationCallback.call(&termParams);
+        return BLE_ERROR_NONE;
+    }
+
+    // check if we can run this discovery
+    if (isConnectionInUse(connHandle)) { 
+        return BLE_STACK_BUSY;
+    }
+
+    // get a new discovery slot, if none are available, just return
+    Discovery* discovery = getAvailableDiscoverySlot();
+    if(discovery == NULL) {
+        return BLE_STACK_BUSY; 
+    }
+
+    // try to launch the discovery 
+    ble_error_t err = gattc_descriptors_discover(connHandle, descriptorStartHandle, descriptorEndHandle);
+    if(!err) { 
+        // commit the new discovery to its slot
+        *discovery = Discovery(characteristic, discoveryCallback, terminationCallback);
+    }
+
+    return err;
+}
+
+bool nRF5xCharacteristicDescriptorDiscoverer::isActive(const DiscoveredCharacteristic& characteristic) const {
+    return findRunningDiscovery(characteristic) != NULL;
+}
+
+void nRF5xCharacteristicDescriptorDiscoverer::requestTerminate(const DiscoveredCharacteristic& characteristic) {
+    Discovery* discovery = findRunningDiscovery(characteristic);
+    if(discovery) { 
+        discovery->onDiscovery = emptyDiscoveryCallback;
+        // call terminate anyway
+        discovery->terminate(BLE_ERROR_NONE);
+        discovery->onTerminate = emptyTerminationCallback;
```

Comment:
Yes, this is the request to terminate the discovery process, it does not mean that anything is done at this point. The termination will occur from the stack side (see the function `terminate` or `process`) and will call the terminate member of the discovery. 

The `onTerminate` member shall better be something valid, replacing the user onTerminate function by an empty one seems the right choice for me. 

Otherwise it is possible to add special member flag indicating if the discovery is terminated from the user side and test if this flag is set every time we try to call terminate but I have the feeling that it makes the code less elegant, less efficient since the condition should always be tested and less understandable since there will be flags test all over the place.